File: implementations/sdk/ruby/lib/lcp_sdk.rb
# frozen_string_literal: true

require "json"
require "openssl"
require "securerandom"
require "time"
require "net/http"
require "uri"
require_relative "schema_validator"
require_relative "generated_models"

module LcpSdk
  Message = Struct.new(:id, :type, :timestamp, :sender_id, :receiver_id, :correlation_id, :idempotency_key, :test, keyword_init: true)

  module Signing
    module_function

    def canonical_bytes(timestamp, idempotency_key, body)
      "#{timestamp}\n#{idempotency_key || ""}\n".b + body.b
    end

    def sign(secret, timestamp, idempotency_key, body)
      OpenSSL::HMAC.hexdigest("SHA256", secret, canonical_bytes(timestamp, idempotency_key, body))
    end

    def verify(secret, signature, timestamp, idempotency_key, body, max_skew_seconds: 300, now: Time.now.utc)
      signed_at = Time.iso8601(timestamp)
      raise ArgumentError, "timestamp outside replay window" if (now - signed_at).abs > max_skew_seconds
      expected = sign(secret, timestamp, idempotency_key, body)
      actual = signature.downcase
      raise ArgumentError, "invalid signature" unless expected.bytesize == actual.bytesize && secure_equal?(expected, actual)
      true
    end

    def secure_equal?(left, right)
      difference = 0
      left.bytes.zip(right.bytes) { |a, b| difference |= a ^ b }
      difference.zero?
    end

    def verify_request(secret, headers, raw_body, max_skew_seconds: 300, now: Time.now.utc)
      normalized = headers.to_h.transform_keys(&:downcase)
      required = %w[x-lcp-signature x-lcp-timestamp x-lcp-idempotency-key]
      missing = required.reject { |key| normalized[key] && !normalized[key].empty? }
      raise ArgumentError, "missing required LCP headers: #{missing.join(", ")}" unless missing.empty?
      verify(secret, normalized["x-lcp-signature"], normalized["x-lcp-timestamp"], normalized["x-lcp-idempotency-key"], raw_body, max_skew_seconds: max_skew_seconds, now: now)
    end
  end

  module_function

  def build_envelope(type, sender_id, receiver_id, payload, test: false)
    id = SecureRandom.uuid
    { "lcp" => { "version" => "1.0.0", "message" => { "id" => id, "type" => type, "timestamp" => Time.now.utc.strftime("%Y-%m-%dT%H:%M:%SZ"), "sender_id" => sender_id, "receiver_id" => receiver_id, "correlation_id" => nil, "idempotency_key" => "#{sender_id}-#{type}-#{SecureRandom.hex(16)}", "test" => test }, "payload" => payload } }
  end

  def validate_envelope(envelope)
    message = envelope.dig("lcp", "message")
    raise ArgumentError, "invalid LCP envelope" unless envelope.dig("lcp", "version").is_a?(String) && message.is_a?(Hash)
    %w[id type timestamp sender_id receiver_id idempotency_key].each { |key| raise ArgumentError, "missing lcp.message.#{key}" unless message[key].is_a?(String) && !message[key].empty? }
    raise ArgumentError, "missing lcp.payload" unless envelope.dig("lcp").key?("payload")
    true
  end

  class Client
    def initialize(endpoint, sender_id: nil, api_key: nil, hmac_secret: nil, max_retries: 2)
      @endpoint = endpoint.delete_suffix("/"); @sender_id = sender_id; @api_key = api_key; @hmac_secret = hmac_secret; @max_retries = max_retries
    end

    def request(method, path, payload: nil, test: false)
      body = payload.nil? ? "" : JSON.generate(payload)
      key = payload&.dig("lcp", "message", "idempotency_key")
      (0..@max_retries).each do |attempt|
        uri = URI.join("#{@endpoint}/", path.delete_prefix("/")); request = Net::HTTP.const_get(method.capitalize).new(uri)
        request["Content-Type"] = "application/json"; request["Accept"] = "application/json"; request["X-LCP-Sender-Id"] = @sender_id if @sender_id; request["Authorization"] = "Bearer #{@api_key}" if @api_key; request["X-LCP-Idempotency-Key"] = key if key; request["X-LCP-Test"] = "true" if test
        if @hmac_secret; timestamp = Time.now.utc.strftime("%Y-%m-%dT%H:%M:%SZ"); request["X-LCP-Timestamp"] = timestamp; request["X-LCP-Signature"] = Signing.sign(@hmac_secret, timestamp, key, body); end
        request.body = body unless body.empty?
        response = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |http| http.request(request) }
        data = response.body.nil? || response.body.empty? ? {} : JSON.parse(response.body)
        return data if response.is_a?(Net::HTTPSuccess)
        raise "LCP HTTP #{response.code}: #{data}" unless [429, 500, 502, 503, 504].include?(response.code.to_i) && attempt < @max_retries
        sleep(2**attempt)
      end
      raise "LCP request failed"
    end

    def submit_lead(envelope, test: false); LcpSdk.validate_envelope(envelope); request("POST", "/v1/lcp/leads", payload: envelope, test: test); end
    def submit_call(envelope, test: false); LcpSdk.validate_envelope(envelope); request("POST", "/v1/lcp/calls", payload: envelope, test: test); end
    def submit_bid(envelope, test: false); LcpSdk.validate_envelope(envelope); request("POST", "/v1/lcp/bids", payload: envelope, test: test); end
    def query_lead_status(id); request("GET", "/v1/lcp/leads/#{URI.encode_www_form_component(id)}"); end
    def get_capabilities; request("GET", "/v1/lcp/capabilities"); end
    def list_offers(vertical = nil); request("GET", "/v1/lcp/offers#{vertical ? "?vertical=#{URI.encode_www_form_component(vertical)}" : ""}"); end
  end
end

File: implementations/sdk/php/src/LcpSdk.php
<?php

declare(strict_types=1);

namespace Lcp;

require_once __DIR__ . '/SchemaValidator.php';
require_once __DIR__ . '/GeneratedModels.php';

use DateTimeImmutable;
use DateTimeZone;
use RuntimeException;

final class Signing
{
    public static function canonicalBytes(string $timestamp, ?string $idempotencyKey, string $body): string
    {
        return $timestamp . "\n" . ($idempotencyKey ?? '') . "\n" . $body;
    }

    public static function sign(string $secret, string $timestamp, ?string $idempotencyKey, string $body): string
    {
        return hash_hmac('sha256', self::canonicalBytes($timestamp, $idempotencyKey, $body), $secret);
    }

    public static function verify(string $secret, string $signature, string $timestamp, ?string $idempotencyKey, string $body, int $maxSkewSeconds = 300, ?DateTimeImmutable $now = null): void
    {
        try { $signedAt = new DateTimeImmutable($timestamp); } catch (\Exception $error) { throw new RuntimeException('Invalid X-LCP-Timestamp', 0, $error); }
        $current = $now ?? new DateTimeImmutable('now', new DateTimeZone('UTC'));
        if (abs($current->getTimestamp() - $signedAt->getTimestamp()) > $maxSkewSeconds) throw new RuntimeException('X-LCP-Timestamp is outside the replay window');
        if (!hash_equals(self::sign($secret, $timestamp, $idempotencyKey, $body), strtolower($signature))) throw new RuntimeException('Invalid X-LCP-Signature');
    }

    public static function verifyRequest(string $secret, array $headers, string $rawBody, int $maxSkewSeconds = 300): void
    {
        $normalized = [];
        foreach ($headers as $key => $value) $normalized[strtolower($key)] = (string)$value;
        foreach (['x-lcp-signature', 'x-lcp-timestamp', 'x-lcp-idempotency-key'] as $required) if (empty($normalized[$required])) throw new RuntimeException("Missing required LCP header: $required");
        self::verify($secret, $normalized['x-lcp-signature'], $normalized['x-lcp-timestamp'], $normalized['x-lcp-idempotency-key'], $rawBody, $maxSkewSeconds);
    }
}

final class Envelope
{
    public static function build(string $type, string $senderId, string $receiverId, array $payload, bool $test = false): array
    {
        $id = self::uuid();
        return ['lcp' => ['version' => '1.0.0', 'message' => ['id' => $id, 'type' => $type, 'timestamp' => gmdate('Y-m-d\\TH:i:s\\Z'), 'sender_id' => $senderId, 'receiver_id' => $receiverId, 'correlation_id' => null, 'idempotency_key' => "$senderId-$type-" . self::uuid(), 'test' => $test], 'payload' => $payload]];
    }

    public static function validate(array $envelope): void
    {
        $message = $envelope['lcp']['message'] ?? null;
        foreach (['version' => $envelope['lcp']['version'] ?? null, 'id' => $message['id'] ?? null, 'type' => $message['type'] ?? null, 'timestamp' => $message['timestamp'] ?? null, 'sender_id' => $message['sender_id'] ?? null, 'receiver_id' => $message['receiver_id'] ?? null, 'idempotency_key' => $message['idempotency_key'] ?? null] as $field => $value) if (!is_string($value) || $value === '') throw new RuntimeException("Missing LCP field: $field");
        if (!array_key_exists('payload', $envelope['lcp'])) throw new RuntimeException('Missing lcp.payload');
    }

    private static function uuid(): string
    {
        $bytes = random_bytes(16); $bytes[6] = chr((ord($bytes[6]) & 0x0f) | 0x40); $bytes[8] = chr((ord($bytes[8]) & 0x3f) | 0x80);
        return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($bytes), 4));
    }
}

final class Client
{
    public function __construct(private readonly string $endpoint, private readonly ?string $senderId = null, private readonly ?string $apiKey = null, private readonly ?string $hmacSecret = null, private readonly int $maxRetries = 2) {}

    public function request(string $method, string $path, ?array $payload = null, bool $test = false): array
    {
        $body = $payload === null ? '' : json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES | JSON_THROW_ON_ERROR);
        $key = $payload['lcp']['message']['idempotency_key'] ?? null;
        for ($attempt = 0; $attempt <= max(0, $this->maxRetries); $attempt++) {
            $headers = ['Content-Type: application/json', 'Accept: application/json']; if ($this->senderId) $headers[] = 'X-LCP-Sender-Id: ' . $this->senderId; if ($this->apiKey) $headers[] = 'Authorization: Bearer ' . $this->apiKey; if ($key) $headers[] = 'X-LCP-Idempotency-Key: ' . $key; if ($test) $headers[] = 'X-LCP-Test: true';
            if ($this->hmacSecret) { $timestamp = gmdate('Y-m-d\\TH:i:s\\Z'); $headers[] = 'X-LCP-Timestamp: ' . $timestamp; $headers[] = 'X-LCP-Signature: ' . Signing::sign($this->hmacSecret, $timestamp, $key, $body); }
            $handle = curl_init(rtrim($this->endpoint, '/') . '/' . ltrim($path, '/')); curl_setopt_array($handle, [CURLOPT_CUSTOMREQUEST => $method, CURLOPT_POSTFIELDS => $body === '' ? null : $body, CURLOPT_HTTPHEADER => $headers, CURLOPT_RETURNTRANSFER => true, CURLOPT_TIMEOUT => 30]); $responseBody = curl_exec($handle); $error = curl_error($handle); $status = (int)curl_getinfo($handle, CURLINFO_RESPONSE_CODE); curl_close($handle);
            if ($error !== '' && $attempt < $this->maxRetries) { sleep(2 ** $attempt); continue; } if ($error !== '') throw new RuntimeException($error);
            $result = json_decode((string)$responseBody, true) ?? ['raw' => $responseBody]; if ($status >= 200 && $status < 300) return $result; if (!in_array($status, [429, 500, 502, 503, 504], true) || $attempt === $this->maxRetries) throw new RuntimeException("LCP HTTP $status: " . json_encode($result)); sleep(2 ** $attempt);
        }
        throw new RuntimeException('LCP request failed');
    }

    public function submitLead(array $envelope, bool $test = false): array { Envelope::validate($envelope); return $this->request('POST', '/v1/lcp/leads', $envelope, $test); }
    public function submitCall(array $envelope, bool $test = false): array { Envelope::validate($envelope); return $this->request('POST', '/v1/lcp/calls', $envelope, $test); }
    public function submitBid(array $envelope, bool $test = false): array { Envelope::validate($envelope); return $this->request('POST', '/v1/lcp/bids', $envelope, $test); }
    public function queryLeadStatus(string $id): array { return $this->request('GET', '/v1/lcp/leads/' . rawurlencode($id)); }
    public function getCapabilities(): array { return $this->request('GET', '/v1/lcp/capabilities'); }
    public function listOffers(?string $vertical = null): array { return $this->request('GET', '/v1/lcp/offers' . ($vertical === null ? '' : '?vertical=' . rawurlencode($vertical))); }
}
